lib/linked_list.rb

```diff
@@ -18,128 +18,299 @@ def initialize
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), n = length of linked list
+    # Space Complexity: O(1)
     def reverse
-      raise NotImplementedError
+      # assign three pointers: one to head, one to head.next, one to head.next.next
+      p1 = @head
+      p2 = p1.next
+      p3 = p2.next
+
+      # tell head to point to nil
+      p1.next = nil
+
+      # point p2 to p1 and move all pointers forward
+      while p3
+        p2.next = p1
+        p1 = p2
+        p2 = p3
+        p3 = p3.next
+      end
+
+      # last link
+      p2.next = p1
+    
+      # once p3 is nil, p2 is pointing at the (former) tail, now head
+      @head = p2
     end
 
     # method that returns the value of the last node in the linked list
     # returns nil if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), n = length of linked list
+    # Space Complexity: O(1)
     def get_last
-      raise NotImplementedError
+      return nil if @head.nil?
+      
+      current = @head
+      while current.next
+        current = current.next
+      end
+
+      return current.data
     end
   
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), n = length of linked list
+    # Space Complexity: O(1)
     def find_middle_value
-      raise NotImplementedError
+      # two pointers, one moving twice every iteration, the other moving once
+      slow = @head
+      fast = @head
+
+      while fast.next
+        fast = fast.next
+        slow = slow.next
+        if fast.next
+          fast = fast.next
+        else
+          return slow.data
+        end
+      end
+
+      return slow.data
     end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), n = length of linked list
+    # Space Complexity: O(1)
     def find_nth_from_end(n)
-      raise NotImplementedError
+      count = 0
+      lead = @head
+      tail = @head
+
+      until count == n
+        return nil if lead.nil?
+        lead = lead.next
+        count += 1
+      end
+
+      return nil if lead.nil?
+
+      until lead.next.nil?
+        tail = tail.next
+        lead = lead.next
+      end
+
+      return tail.data
     end
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), n = length of linked list
+    # Space Complexity: O(n), n = number of nodes (length of linked list)
     def has_cycle

```

Comment:
That is one solution.  However since you are using `.include?` this solution is O(n^2) in time complexity.

You can do it in O(n) with a fast and slow pointer.